user-operator/src/main/java/io/strimzi/operator/user/operator/QuotasOperator.java

```diff
@@ -122,44 +169,43 @@ protected Future<ReconcileResult<KafkaUserQuotas>> internalAlter(Reconciliation
      * @param reconciliation The reconciliation
      * @param username Name of the user
      *
-     * @return the Future with reconcile result
+     * @return the CompletionStage with reconcile result
      */
-    protected Future<KafkaUserQuotas> getAsync(Reconciliation reconciliation, String username) {
-        ClientQuotaFilterComponent c = ClientQuotaFilterComponent.ofEntity(ClientQuotaEntity.USER, username);
-        ClientQuotaFilter f =  ClientQuotaFilter.contains(List.of(c));
-
-        return Util.kafkaFutureToVertxFuture(reconciliation, vertx, adminClient.describeClientQuotas(f).entities())
-                .compose(quotas -> {
-                    ClientQuotaEntity cqe = new ClientQuotaEntity(Map.of(ClientQuotaEntity.USER, username));
-                    KafkaUserQuotas current = null;
+    private CompletionStage<KafkaUserQuotas> getAsync(Reconciliation reconciliation, String username) {
+        LOGGER.debugCr(reconciliation, "Getting quotas for user {}", username);
+        return CompletableFuture.supplyAsync(() -> cache.get(username));
+    }
 
-                    if (quotas.containsKey(cqe)) {
-                        current = QuotaUtils.fromClientQuota(quotas.get(cqe));
-                    }
+    /**
+     * Utility method to generate object with null quotas
+     *
+     * @return  KafkaUserQuotas with all quotas being null
+     */
+    private static KafkaUserQuotas emptyQuotas()   {
+        KafkaUserQuotas emptyQuotas = new KafkaUserQuotas();
+        emptyQuotas.setProducerByteRate(null);
+        emptyQuotas.setConsumerByteRate(null);
+        emptyQuotas.setRequestPercentage(null);
+        emptyQuotas.setControllerMutationRate(null);
 
-                    return Future.succeededFuture(current);
-                });
+        return emptyQuotas;
     }
 
     /**
      * @return Set with all usernames which have some ACLs set
      */
     @Override
-    public Future<Set<String>> getAllUsers() {
+    public CompletionStage<Set<String>> getAllUsers() {
         LOGGER.debugOp("Searching for Users with any quotas");
+        return CompletableFuture.supplyAsync(() -> {
```

Comment:
I was wondering if this method really needs to submit the task asynchronously to run on a different thread. What I see it is just getting data from an already filled cache and filling a Set. Do you expect it could take long time, for this reason it's better to be async? Otherwise doing the process and returning a `CompletableFuture.completedFuture(users)` could be enough? Maybe the same could apply to other operators who access the cache?

Reply:
You are right. This should not be needed. I kept the async interface because the _all usernames_ are also supplied by Kubernetes from a blocking operation. But here using `CompletableFuture.completedFuture(users)` should be enough so I changed it.